src/crypto/store/indexeddb-crypto-store.js

```diff
@@ -110,6 +111,238 @@ export default class IndexedDBCryptoStore {
+                        `not sending another`,
+                    );
+                    deferred.resolve(existing);
+                },
+                onNotFound: () => {
+                    // we got to the end of the list without finding a match
+                    // - add the new request.
+                    console.log(
+                        `enqueueing key request for ${requestBody.room_id} / ` +
+                        requestBody.session_id,
+                    );
+                    const store = txn.objectStore("outgoingRoomKeyRequests");
+                    store.add(request);
+                    txn.onsuccess = () => { deferred.resolve(request); };
+                },
+            });
+
+            return deferred.promise;
+        });
+    }
+
+    /**
+     * Look for an existing room key request
+     *
+     * @param {module:crypto~RoomKeyRequestBody} requestBody
+     *    existing request to look for
+     *
+     * @return {Promise} resolves to the matching
+     *    {@link module:crypto/store/base~OutgoingRoomKeyRequest}, or null if
+     *    not found
+     */
+    getOutgoingRoomKeyRequest(requestBody) {
+        return this.connect().then((db) => {
+            const deferred = q.defer();
+
+            const txn = db.transaction("outgoingRoomKeyRequests", "readonly");
+            txn.onerror = deferred.reject;
+
+            this._getOutgoingRoomKeyRequest(txn, requestBody, {
+                onFound: (existing) => {
+                    // this entry matches the request - return it.
+                    deferred.resolve(existing);
+                },
+                onNotFound: () => {
+                    deferred.resolve(null);
+                },
+            });
+            return deferred.promise;
+        });
+    }
+
+    _getOutgoingRoomKeyRequest(txn, requestBody, opts) {
+        opts = opts || {};
+        const store = txn.objectStore("outgoingRoomKeyRequests");
+
+        const idx = store.index("session");
+        const cursorReq = idx.openCursor([
+            requestBody.room_id,
+            requestBody.session_id,
+        ]);
+
+        cursorReq.onsuccess = (ev) => {
+            const cursor = ev.target.result;
+            if(!cursor) {
+                if (opts.onNotFound) { opts.onNotFound(); }
+                return;
+            }
+
+            const existing = cursor.value;
+
+            if (utils.deepCompare(existing.requestBody, requestBody)) {
+                if (opts.onFound) { opts.onFound(existing); }
+                return;
+            }
+            cursor.continue();
+        };
+    }
+
+    /**
+     * Look for room key requests by state
+     *
+     * @param {Array<Number>} wantedStates list of acceptable states
+     *
+     * @return {Promise} resolves to the a
+     *    {@link module:crypto/store/base~OutgoingRoomKeyRequest}, or null if
+     *    there are no pending requests in those states
```

Comment:
As per your comment on https://github.com/matrix-org/matrix-js-sdk/pull/448/files#r119830241 - I would mention here that if there are multiple matching requests, an arbitrary request is chosen.